src/Features/CSharp/Portable/RemoveUnusedParametersAndValues/CSharpRemoveUnusedParametersAndValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,187 @@
+            if (firstStatement != null)
+            {
+                editor.InsertBefore(firstStatement, declarationStatement);
+            }
+            else
+            {
+                // Switch section without any statements is an error case.
+                // Insert before containing switch statement.
+                editor.InsertBefore(switchCaseBlock.Parent, declarationStatement);
+            }
+        }
+
+        protected override Task<SyntaxNode> RemoveDiscardDeclarationsAsync(
+            SyntaxNode memberDeclaration,
+            Document document,
+            CancellationToken cancellationToken)
+        {
+            var editor = new SyntaxEditor(memberDeclaration, SyntaxGenerator.GetGenerator(document));
+            foreach (var child in memberDeclaration.DescendantNodes(descendIntoChildren: n => !(n is ExpressionSyntax)))
+            {
+                if (child is LocalDeclarationStatementSyntax localDeclarationStatement &&
+                    localDeclarationStatement.Declaration.Variables.Any(v => v.Identifier.Text == "_"))
+                {
+                    ProcessVariableDeclarationWithDiscard(localDeclarationStatement, editor);
+                }
+                else if (child is CatchDeclarationSyntax catchDeclaration &&
+                    catchDeclaration.Identifier.Text == "_")
+                {
+                    // "catch (Exception _)" => "catch (Exception)"
+                    var newCatchDeclaration = catchDeclaration.WithIdentifier(default)
+                                              .WithAdditionalAnnotations(Formatter.Annotation);
+                    editor.ReplaceNode(catchDeclaration, newCatchDeclaration);
+                }
+            }
+
+            return Task.FromResult(editor.GetChangedRoot());
+        }
+
+        private static void ProcessVariableDeclarationWithDiscard(
+            LocalDeclarationStatementSyntax localDeclarationStatement,
+            SyntaxEditor editor)
+        {
+            var statementsBuilder = ArrayBuilder<StatementSyntax>.GetInstance();
+            var variableDeclaration = localDeclarationStatement.Declaration;
+            var currentNonDiscardVariables = new SeparatedSyntaxList<VariableDeclaratorSyntax>();
+
+            try
+            {
+                foreach (var variable in variableDeclaration.Variables)
+                {
+                    if (variable.Identifier.Text != "_")
+                    {
+                        currentNonDiscardVariables = currentNonDiscardVariables.Add(variable);
+                    }
+                    else
+                    {
+                        ProcessCurrentNonDiscardVariables();
+                        ProcessDiscardVariable(variable);
+                    }
+                }
+
+                ProcessCurrentNonDiscardVariables();
+
+                if (statementsBuilder.Count == 0)
+                {
+                    return;
+                }
```

Comment:
it appears as if this method is using mutation by capture.  this is really hard to understand while reading.  From the reading up till now, statementsBuilder has never been touched.  Making it seem impossible for this to be anything other htan 0.  If you are going to mutate captured state, instead just pass in the variables to be mutated to the local functions.  This makes data flow very easy to reason about and helps inform the user whihc of the above method calls are actually mutating state.

Reply:
Refactored the code for improved readability and added explanatory comments.

---
In reply to: [230603456](https://github.com/dotnet/roslyn/pull/30777#discussion_r230603456) [](ancestors = 230603456)